SwiftMQTT/SwiftMQTT/MQTTReconnectingSession.swift

```diff
@@ -0,0 +1,151 @@
+    func mqttDidReceive(message: MQTTMessage, from session: MQTTReconnectingSession)
+    func mqttConnected(_ changed: Bool, for: MQTTReconnectingSession, error: Error?)
+}
+
+/*
+    TODO: encapsulate credentials
+*/
+
+public struct MQTTCredentials {
+    // username, password, useSSL, certs, lastWill
+}
+
+public struct MQTTConnectParams {
+    public init(clientID: String) {
+        self.clientID = clientID
+    }
+    public var clientID: String
+    public var host: String = "localhost"
+    public var port: UInt16 = 1883
+    public var cleanSession: Bool = true
+    public var keepAlive: UInt16 = 15
+    public var useSSL: Bool = false
+	
+    public var timeout: TimeInterval = 1.0
+    public var retryCount: Int = 3
+    public var retryTimeInterval: TimeInterval = 1.0
+    public var resuscitateTimeInterval: TimeInterval = 5.0
+}
+
+public extension MQTTSession {
+    public convenience init(connectParams: MQTTConnectParams) {
+        self.init(
+			host: connectParams.host,
+			port: connectParams.port,
+			clientID: connectParams.clientID,
+			cleanSession: connectParams.cleanSession,
+			keepAlive: connectParams.keepAlive,
+			connectionTimeout: connectParams.timeout,
+			useSSL: connectParams.useSSL)
+    }
+}
+
+public class MQTTReconnectingSession: MQTTBroker {
+    fileprivate let connectParams: MQTTConnectParams
+    fileprivate let batchPredicate: ([MQTTMessage])->Bool
+    fileprivate var session: MQTTSession!
+    
+    open weak var delegate: MQTTReconnectingSessionDelegate?
+    
+    public init(connectParams: MQTTConnectParams, batchPredicate: @escaping ([MQTTMessage])->Bool) {
+        self.batchPredicate = batchPredicate
+        self.connectParams = connectParams
+    }
+    
+    public func connect(completion: MQTTSessionCompletionBlock? = nil) {
+        self.session = MQTTSession(connectParams: connectParams)
+        self.session.delegate = self
+        self.connectWithRetry(0, nil, completion)
+    }
+    
+    public func publish(_ data: Data, in topic: String, delivering qos: MQTTQoS, retain: Bool, completion: MQTTSessionCompletionBlock?) {
+        self.session.publish(data, in: topic, delivering: qos, retain: retain, completion: completion)
+    }
+    
+    public func subscribe(to topics: [String: MQTTQoS], completion: MQTTSessionCompletionBlock?) {
+        self.session.subscribe(to: topics, completion: completion)
+    }
+    
+    public func unSubscribe(from topics: [String], completion: MQTTSessionCompletionBlock?) {
+        self.session.unSubscribe(from: topics, completion: completion)
+    }
+    
+    public func disconnect() {
+        self.session = nil
+		self.delegate?.mqttConnected(false, for: self, error: nil)
+    }
+}
+
+extension MQTTReconnectingSession {
+    fileprivate func connectWithRetry(_ attempt: Int, _ error: Error?, _ completion: MQTTSessionCompletionBlock?) {
+        self.session.connect { [weak self] success, newError in
```

Comment:
Not related to this PR, but, `Result` enum is a perfect candidate for this case. We can remove `success` and `newError` and use `Result` cases instead.